rust/worker/src/index/fulltext/types.rs

```diff
@@ -116,227 +50,605 @@ impl FullTextIndex for BlockfileFullTextIndex {
             // See https://docs.rs/tantivy/latest/tantivy/tokenizer/struct.Token.html
             if !builder.contains_doc_id(offset_id) {
                 // Casting to i32 is safe since we limit the size of the document.
-                builder.add_doc_id_and_positions(offset_id, vec![token.offset_from as i32]);
+                let res =
+                    builder.add_doc_id_and_positions(offset_id, vec![token.offset_from as i32]);
+                if res.is_err() {
+                    return res;
+                }
             } else {
-                builder.add_positions_for_doc_id(offset_id, vec![token.offset_from as i32]);
+                let res =
+                    builder.add_positions_for_doc_id(offset_id, vec![token.offset_from as i32]);
+                if res.is_err() {
+                    return res;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    async fn write_to_blockfiles(&mut self) -> Result<(), Box<dyn ChromaError>> {
+        for (key, mut value) in self.uncommitted.drain() {
+            let built_list = value.build();
+            for doc_id in built_list.doc_ids.iter() {
+                match doc_id {
+                    Some(doc_id) => {
+                        let positional_posting_list =
+                            built_list.get_positions_for_doc_id(doc_id).unwrap();
+                        let res = self
+                            .posting_lists_blockfile_writer
+                            .set(key.as_str(), doc_id as u32, &positional_posting_list)
+                            .await;
+                        if res.is_err() {
+                            return res;
+                        }
+                    }
+                    None => {
+                        panic!("Positions for doc ID not found in positional posting list -- should never happen")
+                    }
+                }
+            }
+        }
+        for (key, value) in self.uncommitted_frequencies.drain() {
+            // TODO we just have token -> frequency here. Should frequency be the key or should we use an empty key and make it the value?
+            let res = self
+                .frequencies_blockfile_writer
+                .set(key.as_str(), value as u32, 0)
+                .await;
+            if res.is_err() {
+                return res;
             }
         }
         Ok(())
     }
 
-    fn search(&mut self, query: &str) -> Result<Vec<i32>, Box<dyn ChromaError>> {
+    async fn commit_and_flush(self) -> Result<(), Box<dyn ChromaError>> {
```

Comment:
I don't think we want this method at the index level, if you look at the higher level pattern we want commit and flush as two steps with a type safe transition